athena-elasticsearch/src/main/java/com/amazonaws/connectors/athena/elasticsearch/ElasticsearchRecordHandler.java

```diff
@@ -0,0 +1,215 @@
+     * @param recordsRequest Details of the read request, including:
+     * 1. The Split
+     * 2. The Catalog, Database, and Table the read request is for.
+     * 3. The filtering predicate (if any)
+     * 4. The columns required for projection.
+     * @param queryStatusChecker A QueryStatusChecker that you can use to stop doing work for a query that has already terminated
+     * @throws IOException
+     * @note Avoid writing >10 rows per-call to BlockSpiller.writeRow(...) because this will limit the BlockSpiller's
+     * ability to control Block size. The resulting increase in Block size may cause failures and reduced performance.
+     */
+    @Override
+    protected void readWithConstraint(BlockSpiller spiller, ReadRecordsRequest recordsRequest,
+                                      QueryStatusChecker queryStatusChecker)
+    {
+        logger.info("readWithConstraint - enter\n\nDomain: {}\n\nIndex: {}\n\nMapping: {}\n\nConstraints: {}",
+                recordsRequest.getTableName().getSchemaName(), recordsRequest.getTableName().getTableName(),
+                recordsRequest.getSchema(), recordsRequest.getConstraints().getSummary());
+
+        String endpoint = ElasticsearchHelper.getDomainEndpoint(recordsRequest.getTableName().getSchemaName());
+        long numRows = 0;
+        AtomicLong numResultRows = new AtomicLong(0);
+
+        if (!endpoint.isEmpty() && queryStatusChecker.isQueryRunning()) {
+            AwsRestHighLevelClient client = clientFactory.getClient(endpoint);
+            try {
+                // Create a new search-source injected with the projection, predicate, and the pagination batch size.
+                SearchSourceBuilder searchSource = new SearchSourceBuilder().size(QUERY_BATCH_SIZE)
+                        .fetchSource(ElasticsearchHelper
+                                .getProjection(recordsRequest.getSchema())).query(ElasticsearchHelper
+                                .getQuery(recordsRequest.getConstraints().getSummary()));
+                // Create a new search-request for the specified index.
+                SearchRequest searchRequest = new SearchRequest(recordsRequest.getTableName().getTableName());
+                int hitsNum;
+                int currPosition = 0;
+                do {
+                    // Process the search request injecting the search-source, and setting the from position
+                    // used for pagination of results.
+                    SearchResponse searchResponse = client
+                            .getDocuments(searchRequest.source(searchSource.from(currPosition)));
+                    // Increment current position to next batch of results.
+                    currPosition += QUERY_BATCH_SIZE;
+                    // Process hits.
+                    Iterator<SearchHit> hitIterator = searchResponse.getHits().iterator();
+                    hitsNum = searchResponse.getHits().getHits().length;
```

Comment:
why GetHits().GetHits()? weird that its two layers.

Reply:
Yes, it is a weird implementation, but it seems to be the best way to get the number of hits (documents) returned from the query. The first GetHits() returns a SearchHits object. The next GetHits() returns a SearchHit[]. The only other way I can think of is to set hitsNum to 0 prior to the processing loop, and then increment it inside the loop for each document processed...